File: packages/llmvault-mobile/android/src/main/java/io/llmvault/mobile/PolicyEnforcer.kt
package io.llmvault.mobile

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

// ---------------------------------------------------------------------------
// Supporting types
// ---------------------------------------------------------------------------

data class AllowlistEntry(
    val hostPattern: String,
    val httpsOnly: Boolean,
)

data class RelayPolicyNative(
    val schemaVersion: Int,
    val providerAllowlist: List<AllowlistEntry>,
    val maxTokensPerRequest: Int,
    val dailyCostLimitMicrounits: Long,
    val monthlyCostLimitMicrounits: Long,
    val monthlyWarningThresholdPct: Int,
    val highCostThresholdMicrounits: Long,
    val biometricAutoApproveSeconds: Int = DEFAULT_BIOMETRIC_AUTO_APPROVE_SECONDS,
    val blockPrivateIps: Boolean,
) {
    companion object {
        /**
         * Default biometric auto-approve window in seconds.
         *
         * Matches the TS default in packages/secure-relay/src/policy.ts and
         * the iOS LAContext reuse cap
         * (LATouchIDAuthenticationMaximumAllowableReuseDuration = 300s).
         * Also aligns with the time-bound Keystore validity window set in
         * SecureKeyStore.ensureMasterKeyExists — so the OS, the iOS cap,
         * and our app-layer timer all agree on the same 5-minute cadence.
         */
        const val DEFAULT_BIOMETRIC_AUTO_APPROVE_SECONDS = 300

        val defaultPolicy = RelayPolicyNative(
            schemaVersion = 1,
            providerAllowlist = listOf(
                AllowlistEntry("api.openai.com", httpsOnly = true),
                AllowlistEntry("api.anthropic.com", httpsOnly = true),
                AllowlistEntry("api.groq.com", httpsOnly = true),
                AllowlistEntry("generativelanguage.googleapis.com", httpsOnly = true),
                AllowlistEntry("api.mistral.ai", httpsOnly = true),
            ),
            maxTokensPerRequest = 4096,
            dailyCostLimitMicrounits = 5_000_000L,
            monthlyCostLimitMicrounits = 50_000_000L,
            monthlyWarningThresholdPct = 80,
            highCostThresholdMicrounits = 500_000L,
            biometricAutoApproveSeconds = DEFAULT_BIOMETRIC_AUTO_APPROVE_SECONDS,
            blockPrivateIps = true,
        )
    }
}

sealed class PolicyValidation {
    object Allowed : PolicyValidation()
    data class Rejected(val reason: String) : PolicyValidation()
}

sealed class BudgetCheck {
    data class WithinBudget(val remaining: Long) : BudgetCheck()
    data class OverBudget(val remaining: Long) : BudgetCheck()
}

// ---------------------------------------------------------------------------
// Singleton enforcer
// ---------------------------------------------------------------------------

/**
 * Native-side policy enforcement (defense in depth — mirrors policy.ts and the Swift PolicyEnforcer).
 * Validates URLs, enforces daily/monthly budgets, and determines biometric requirements.
 *
 * Singleton: access via [PolicyEnforcer] directly (object declaration).
 * All public methods are @Synchronized to guard accumulated-cost state on the JVM.
 */
object PolicyEnforcer {

    // Current policy — can be replaced at runtime via [updatePolicy].
    @Volatile
    var policy: RelayPolicyNative = RelayPolicyNative.defaultPolicy
        private set

    // Daily budget state
    @Volatile
    var dailyAccumulated: Long = 0L
        private set
    private var dailyResetDate: String = ""

    // Monthly budget state
    @Volatile
    var monthlyAccumulated: Long = 0L
        private set
    private var monthlyResetMonth: String = ""

    // -----------------------------------------------------------------------
    // Private IP detection
    // -----------------------------------------------------------------------

    private val privateIpPatterns: List<Regex> = listOf(
        Regex("^10\\."),
        Regex("^172\\.(1[6-9]|2\\d|3[01])\\."),
        Regex("^192\\.168\\."),
        Regex("^127\\."),
        Regex("^0\\."),
        Regex("^169\\.254\\."),
        Regex("^::1$"),
        Regex("^fd"),
        Regex("^fc"),
        Regex("^fe80:"),
    )

    private fun isPrivateIp(host: String): Boolean =
        privateIpPatterns.any { it.containsMatchIn(host) }

    // -----------------------------------------------------------------------
    // Host matching
    // -----------------------------------------------------------------------

    /**
     * Supports exact host matches and wildcard patterns of the form "*.example.com".
     * A wildcard pattern also matches the bare root "example.com".
     */
    private fun matchesHost(pattern: String, host: String): Boolean {
        if (pattern.startsWith("*.")) {
            val suffix = pattern.removePrefix("*") // ".example.com"
            return host.endsWith(suffix) || host == pattern.removePrefix("*.")
        }
        return host == pattern
    }

    // -----------------------------------------------------------------------
    // URL validation
    // -----------------------------------------------------------------------

    /**
     * Validate that [urlString] is allowed by the current policy.
     * Returns [PolicyValidation.Allowed] or [PolicyValidation.Rejected] with a reason.
     */
    fun validateBaseUrl(urlString: String): PolicyValidation {
        val url = runCatching { java.net.URL(urlString) }.getOrNull()
            ?: return PolicyValidation.Rejected("Invalid URL")

        val scheme = url.protocol ?: return PolicyValidation.Rejected("Missing URL scheme")
        val host = url.host ?: return PolicyValidation.Rejected("Missing URL host")

        if (scheme != "https") {
            return PolicyValidation.Rejected("HTTP not allowed; HTTPS required")
        }

        if (policy.blockPrivateIps && isPrivateIp(host)) {
            return PolicyValidation.Rejected("Private IP addresses are blocked")
        }

        val allowed = policy.providerAllowlist.any { entry ->
            matchesHost(entry.hostPattern, host)
        }

        return if (allowed) {
            PolicyValidation.Allowed
        } else {
            PolicyValidation.Rejected("Host $host not in provider allowlist")
        }
    }

    // -----------------------------------------------------------------------
    // Budget checks
    // -----------------------------------------------------------------------

    @Synchronized
    fun checkDailyBudget(estimatedCost: Long): BudgetCheck {
        resetIfNewDay()
        val remaining = policy.dailyCostLimitMicrounits - dailyAccumulated
        return if (estimatedCost > remaining) {
            BudgetCheck.OverBudget(remaining = remaining)
        } else {
            BudgetCheck.WithinBudget(remaining = remaining - estimatedCost)
        }
    }

    @Synchronized
    fun checkMonthlyBudget(estimatedCost: Long): BudgetCheck {
        resetIfNewMonth()
        val remaining = policy.monthlyCostLimitMicrounits - monthlyAccumulated
        return if (estimatedCost > remaining) {
            BudgetCheck.OverBudget(remaining = remaining)
        } else {
            BudgetCheck.WithinBudget(remaining = remaining - estimatedCost)
        }
    }

    /** Record the actual cost of a completed request (in microunits). */
    @Synchronized
    fun recordCost(microunits: Long) {
        resetIfNewDay()
        resetIfNewMonth()
        dailyAccumulated += microunits
        monthlyAccumulated += microunits
    }

    // -----------------------------------------------------------------------
    // Cost estimation (companion-style statics via companion object)
    // -----------------------------------------------------------------------

    /**
     * Estimate cost in microunits for a request.
     * Rates are per-1000-tokens in microunits (1 USD = 1,000,000 microunits).
     */
    fun estimateCost(model: String, promptTokens: Int, maxCompletionTokens: Int): Long {
        val rates = costRates(model)
        val promptCost = promptTokens.toLong() * rates.first / 1000L
        val completionCost = maxCompletionTokens.toLong() * rates.second / 1000L
        return promptCost + completionCost
    }

    /**
     * Returns (promptRatePerKToken, completionRatePerKToken) in microunits.
     * Defaults to claude-sonnet-4 rates when the model is unrecognised.
     */
    private fun costRates(model: String): Pair<Long, Long> = when (model) {
        "gpt-4o"                      -> 2_500L to 10_000L
        "gpt-4o-mini"                 -> 150L to 600L
        "gpt-4-turbo"                 -> 10_000L to 30_000L
        "claude-sonnet-4-20250514"    -> 3_000L to 15_000L
        "claude-haiku-4-20250414"     -> 800L to 4_000L
        else                          -> 3_000L to 15_000L  // conservative default
    }

    // -----------------------------------------------------------------------
    // Biometric requirement
    // -----------------------------------------------------------------------

    /**
     * Returns true if biometric authentication is required before this request
     * may proceed.
     *
     * @param estimatedCost Cost estimate in microunits.
     * @param lastAuthAt    Timestamp of the last successful biometric authentication,
     *                      or null if never authenticated.
     */
    fun requiresBiometric(estimatedCost: Long, lastAuthAt: Date?): Boolean {
        // High-cost requests always require biometric regardless of cached auth.
        if (estimatedCost >= policy.highCostThresholdMicrounits) return true

        // Within the auto-approve window no re-authentication is needed.
        if (lastAuthAt != null) {
            val elapsedSeconds = (Date().time - lastAuthAt.time) / 1000L
            if (elapsedSeconds <= policy.biometricAutoApproveSeconds.toLong()) return false
        }

        return true
    }

    // -----------------------------------------------------------------------
    // Policy update
    // -----------------------------------------------------------------------

    /**
     * Update budget / threshold settings.
     * `providerAllowlist` and `blockPrivateIps` are security-critical and
     * cannot be overridden from the JS bridge — they are always preserved
     * from the current (or default) policy.
     */
    fun updatePolicy(newPolicy: RelayPolicyNative) {
        policy = newPolicy.copy(
            providerAllowlist = policy.providerAllowlist,
            blockPrivateIps = true,
        )
    }

    // -----------------------------------------------------------------------
    // Private date helpers
    // -----------------------------------------------------------------------

    // NOTE: SimpleDateFormat is not thread-safe; allocate per call.

    private fun currentDateString(): String =
        SimpleDateFormat("yyyy-MM-dd", Locale.US).format(Date())

    private fun currentMonthString(): String =
        SimpleDateFormat("yyyy-MM", Locale.US).format(Date())

    /** Must be called inside a @Synchronized block. */
    private fun resetIfNewDay() {
        val today = currentDateString()
        if (today != dailyResetDate) {
            dailyAccumulated = 0L
            dailyResetDate = today
        }
    }

    /** Must be called inside a @Synchronized block. */
    private fun resetIfNewMonth() {
        val month = currentMonthString()
        if (month != monthlyResetMonth) {
            monthlyAccumulated = 0L
            monthlyResetMonth = month
        }
    }
}

File: packages/llmvault-mobile/ios/Sources/SecureRelayPlugin/PolicyEnforcer.swift
import Foundation

/// Native-side policy enforcement (defense in depth — same rules as policy.ts).
/// Validates URLs, checks budgets, and determines biometric requirements.
final class PolicyEnforcer {

    static let shared = PolicyEnforcer()

    // MARK: - Default Policy

    static let defaultAllowlist: [[String: Any]] = [
        ["hostPattern": "api.openai.com", "httpsOnly": true],
        ["hostPattern": "api.anthropic.com", "httpsOnly": true],
        ["hostPattern": "api.groq.com", "httpsOnly": true],
        ["hostPattern": "generativelanguage.googleapis.com", "httpsOnly": true],
        ["hostPattern": "api.mistral.ai", "httpsOnly": true],
    ]

    /// Current stored policy (loaded from Keychain on init).
    private(set) var policy: RelayPolicyNative

    /// Daily accumulated cost in microunits.
    private(set) var dailyAccumulated: Int64 = 0
    private var dailyResetDate: String = ""

    /// Monthly accumulated cost in microunits.
    private(set) var monthlyAccumulated: Int64 = 0
    private var monthlyResetMonth: String = ""

    private init() {
        self.policy = RelayPolicyNative.defaultPolicy
    }

    // MARK: - URL Validation

    /// Validate that a base URL is allowed by the policy.
    func validateBaseUrl(_ urlString: String) -> PolicyValidation {
        guard let url = URL(string: urlString),
              let scheme = url.scheme,
              let host = url.host else {
            return .rejected("Invalid URL")
        }

        // HTTPS only
        guard scheme == "https" else {
            return .rejected("HTTP not allowed; HTTPS required")
        }

        // Block private IPs
        if isPrivateIp(host) {
            return .rejected("Private IP addresses are blocked")
        }

        // Check allowlist
        let allowed = policy.providerAllowlist.contains { entry in
            matchesHost(pattern: entry.hostPattern, host: host)
        }

        return allowed ? .allowed : .rejected("Host \(host) not in provider allowlist")
    }

    // MARK: - Budget Checking

    func checkDailyBudget(estimatedCost: Int64) -> BudgetCheck {
        resetIfNewDay()
        let remaining = policy.dailyCostLimitMicrounits - dailyAccumulated
        if estimatedCost > remaining {
            return .overBudget(remaining: remaining)
        }
        return .withinBudget(remaining: remaining - estimatedCost)
    }

    func checkMonthlyBudget(estimatedCost: Int64) -> BudgetCheck {
        resetIfNewMonth()
        let remaining = policy.monthlyCostLimitMicrounits - monthlyAccumulated
        if estimatedCost > remaining {
            return .overBudget(remaining: remaining)
        }
        return .withinBudget(remaining: remaining - estimatedCost)
    }

    /// Record a completed request cost.
    func recordCost(_ microunits: Int64) {
        resetIfNewDay()
        resetIfNewMonth()
        dailyAccumulated += microunits
        monthlyAccumulated += microunits
    }

    // MARK: - Cost Estimation

    /// Estimate cost in microunits for a request.
    static func estimateCost(model: String, promptTokens: Int, maxCompletionTokens: Int) -> Int64 {
        let rates = costRates(for: model)
        let promptCost = Int64(promptTokens) * rates.prompt / 1000
        let completionCost = Int64(maxCompletionTokens) * rates.completion / 1000
        return promptCost + completionCost
    }

    private static func costRates(for model: String) -> (prompt: Int64, completion: Int64) {
        let costs: [String: (prompt: Int64, completion: Int64)] = [
            "gpt-4o": (2500, 10000),
            "gpt-4o-mini": (150, 600),
            "gpt-4-turbo": (10000, 30000),
            "claude-sonnet-4-20250514": (3000, 15000),
            "claude-haiku-4-20250414": (800, 4000),
        ]
        return costs[model] ?? (3000, 15000) // default
    }

    // MARK: - Biometric Requirements

    /// Check if biometric is required for this request.
    func requiresBiometric(estimatedCost: Int64, lastAuthAt: Date?) -> Bool {
        // High-cost always requires biometric
        if estimatedCost >= policy.highCostThresholdMicrounits {
            return true
        }

        // Within auto-approve window, no biometric needed
        if let lastAuth = lastAuthAt {
            let elapsed = Date().timeIntervalSince(lastAuth)
            if elapsed <= Double(policy.biometricAutoApproveSeconds) {
                return false
            }
        }

        return true
    }

    // MARK: - Policy Update

    /// Update budget / threshold settings.
    /// `providerAllowlist` and `blockPrivateIps` are security-critical and
    /// cannot be overridden from the JS bridge — they are always preserved
    /// from the current (or default) policy.
    func updatePolicy(_ newPolicy: RelayPolicyNative) {
        self.policy = RelayPolicyNative(
            schemaVersion: newPolicy.schemaVersion,
            providerAllowlist: self.policy.providerAllowlist,
            maxTokensPerRequest: newPolicy.maxTokensPerRequest,
            dailyCostLimitMicrounits: newPolicy.dailyCostLimitMicrounits,
            monthlyCostLimitMicrounits: newPolicy.monthlyCostLimitMicrounits,
            monthlyWarningThresholdPct: newPolicy.monthlyWarningThresholdPct,
            highCostThresholdMicrounits: newPolicy.highCostThresholdMicrounits,
            biometricAutoApproveSeconds: newPolicy.biometricAutoApproveSeconds,
            blockPrivateIps: true
        )
    }

    // MARK: - Private Helpers

    private func isPrivateIp(_ host: String) -> Bool {
        let patterns = [
            "^10\\.",
            "^172\\.(1[6-9]|2\\d|3[01])\\.",
            "^192\\.168\\.",
            "^127\\.",
            "^0\\.",
            "^169\\.254\\.",
            "^::1$",
            "^fd",
            "^fc",
            "^fe80:",
        ]
        return patterns.contains { pattern in
            host.range(of: pattern, options: .regularExpression, range: nil, locale: nil) != nil
        }
    }

    private func matchesHost(pattern: String, host: String) -> Bool {
        if pattern.hasPrefix("*.") {
            let suffix = String(pattern.dropFirst(1)) // ".example.com"
            return host.hasSuffix(suffix) || host == String(pattern.dropFirst(2))
        }
        return host == pattern
    }

    private func resetIfNewDay() {
        let today = currentDateString()
        if today != dailyResetDate {
            dailyAccumulated = 0
            dailyResetDate = today
        }
    }

    private func resetIfNewMonth() {
        let month = currentMonthString()
        if month != monthlyResetMonth {
            monthlyAccumulated = 0
            monthlyResetMonth = month
        }
    }

    private func currentDateString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }

    private func currentMonthString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM"
        return formatter.string(from: Date())
    }
}

// MARK: - Types

struct RelayPolicyNative {
    let schemaVersion: Int
    let providerAllowlist: [AllowlistEntry]
    let maxTokensPerRequest: Int
    let dailyCostLimitMicrounits: Int64
    let monthlyCostLimitMicrounits: Int64
    let monthlyWarningThresholdPct: Int
    let highCostThresholdMicrounits: Int64
    let biometricAutoApproveSeconds: Int
    let blockPrivateIps: Bool

    struct AllowlistEntry {
        let hostPattern: String
        let httpsOnly: Bool
    }

    /// Default biometric auto-approve window in seconds.
    ///
    /// Mirrors the TS default in policy.ts and Android's
    /// PolicyEnforcer.DEFAULT_BIOMETRIC_AUTO_APPROVE_SECONDS. 300s is also
    /// the iOS LATouchIDAuthenticationMaximumAllowableReuseDuration hard
    /// cap so the app-layer timer cannot outlive the OS cache.
    static let defaultBiometricAutoApproveSeconds: Int = 300

    static let defaultPolicy = RelayPolicyNative(
        schemaVersion: 1,
        providerAllowlist: [
            AllowlistEntry(hostPattern: "api.openai.com", httpsOnly: true),
            AllowlistEntry(hostPattern: "api.anthropic.com", httpsOnly: true),
            AllowlistEntry(hostPattern: "api.groq.com", httpsOnly: true),
            AllowlistEntry(hostPattern: "generativelanguage.googleapis.com", httpsOnly: true),
            AllowlistEntry(hostPattern: "api.mistral.ai", httpsOnly: true),
        ],
        maxTokensPerRequest: 4096,
        dailyCostLimitMicrounits: 5_000_000,
        monthlyCostLimitMicrounits: 50_000_000,
        monthlyWarningThresholdPct: 80,
        highCostThresholdMicrounits: 500_000,
        biometricAutoApproveSeconds: defaultBiometricAutoApproveSeconds,
        blockPrivateIps: true
    )
}

enum PolicyValidation {
    case allowed
    case rejected(String)
}

enum BudgetCheck {
    case withinBudget(remaining: Int64)
    case overBudget(remaining: Int64)
}
